### ISSUE TITLE
last() gives cryptic error message

### ISSUE THREAD

**Issue by t-kalinowski** [hadley/dplyr#2090]
Hello!

a small example:

```
library(dplyr)
n <- 9
x <- c(ifelse(sample(c(T,F), n, T), 1:n, NA ), NA)
df <-  data_frame(x)

last(na.omit(x)) # works

mutate(df, y = last(na.omit(x)))            # doesn't work
# Error: Unsupported vector type language

mutate(df, z = last(x[!is.na(x)]))          # doesn't work
# Error: Unsupported vector type language

mutate(df, y = x %>% na.omit() %>% last() ) # doesn't work
# Error in mutate_impl(.data, dots) : bad value


mutate(df, y = x %>% na.omit %>% last ) # works
last(x[!is.na(x)]) # works

mutate(df, z = last(x)) # "works"

```

I really love dplyr and am immensely grateful for all the hard work that goes into it! 

I'm not savvy enough to even diagnose what's leading to these error messages, but I suspect this is not intended behavior.

**Comment by the-knife** [comment]
Related to this ?
https://github.com/hadley/dplyr/issues/1980

**Comment by krlmlr** [comment]
Thanks, confirmed. Would you like to contribute a testthat test?

This looks related to, but probably not the same as #1980.

**Comment by holstius** [comment]
Similar behavior exposed via `filter()`.

I would not have thought the two expressions below would yield different outcomes — since `identical(6L, c(6L))` is `TRUE` — but they do:

```
> filter(mtcars, cyl == last(6L)) # FIXME: returns 0 rows
```

```
> filter(mtcars, cyl == last(c(6L))) # FIXME: throws "Unsupported vector type language"
```

In contrast to `last`, `first` yields the expected outcome:

```
# All of these yield the correct result
> filter(mtcars, cyl == first(6L))
> filter(mtcars, cyl == first(c(6L))) 
> filter(mtcars, cyl == 6L)
```

Feels like ... maybe an off-by-one in `nth`?

**Comment by holstius** [comment]
Hi @krlmlr, 

I'm not good at debugging Rcpp stuff — I only know how to insert `stop()` statements, `load_all()`, think, and repeat — but it seems that:

1. Supplying `order_by = NULL` makes the problem disappear. For example, `filter(mtcars, cyl == last(c(6L), order_by = NULL))` produces the correct result, while `filter(mtcars, cyl == last(c(6L)))` throws the error noted by @t-kalinowski ; and

2. The "easy case" (nargs == 2, line 157 of `hybrid_nth.cpp`, `switch(TYPEOF(data))`) only handles `INTSXP`, `REALSXP`, `STRSXP`, and `LGLSXP`. So it falls through when `TYPEOF(data)` is `language`, which happens when `order_by` is not supplied.

Hope these observations are helpful.

**Comment by krlmlr** [comment]
This will be resolved by #2143, but that might take some time until it's ready.

**Comment by BrazilForever11** [comment]
Here is another workaround

`mutate(df,y = tail(na.omit(x),1))`
